addons/point_of_sale/tests/test_pos_stock_account.py

```diff
@@ -235,3 +235,312 @@ def test_03_order_product_w_owner(self):
+
+        # sync orders
+        self.env['pos.order'].create_from_ui(orders)
+
+        # close the session
+        self.pos_session.action_pos_session_validate()
+
+        # picking should be in done state
+        self.assertEqual(self.pos_session.picking_ids[0].state, 'done', 'Picking should be in done state.')
+
+        self._assert_account_move(self.pos_session.move_id, {
+            'line_ids': [
+                {'account_id': self.sales_account.id, 'debit': 0, 'credit': 30, 'reconciled': False},
+                {'account_id': self.expense_account.id, 'debit': 15, 'credit': 0, 'reconciled': False},
+                {'account_id': self.cash_pm1.receivable_account_id.id, 'debit': 30, 'credit': 0, 'reconciled': True},
+                {'account_id': self.output_account.id, 'debit': 0, 'credit': 15, 'reconciled': True},
+            ],
+        })
+
+        # Verify the expected inventory valuation entry was created
+        stock_moves = self.pos_session.picking_ids[0].move_lines.account_move_ids
+        self.assertEqual(len(stock_moves), 1)
+        self._assert_account_move(stock_moves[0], {
+            'line_ids': [
+                {'account_id': self.valuation_account.id, 'debit': 0, 'credit': 15, 'reconciled': False},
+                {'account_id': self.output_account.id, 'debit': 15, 'credit': 0, 'reconciled': True},
+            ],
+        })
+
+    def test_selling_a_product_with_an_invalid_picking(self):
+        self.open_new_session()
+
+        # create orders
+        orders = [self.create_ui_order_data([(self.product1, 1)])]
+
+        # sync orders
+        self.env['pos.order'].create_from_ui(orders)
+
+        # close the session
+        self.pos_session.action_pos_session_validate()
+
+        # picking should be in confirmed state
+        self.assertEqual(self.pos_session.picking_ids[0].state, 'confirmed', 'Picking should be in confirmed state.')
+
+        # The expense and stock output lines will be 0 because the stock picking failed. The stock output line won't
+        # be reconciled.
+        self._assert_account_move(self.pos_session.move_id, {
+            'line_ids': [
+                {'account_id': self.sales_account.id, 'debit': 0, 'credit': 20, 'reconciled': False},
+                {'account_id': self.expense_account.id, 'debit': 0, 'credit': 0, 'reconciled': False},
+                {'account_id': self.cash_pm1.receivable_account_id.id, 'debit': 20, 'credit': 0, 'reconciled': True},
+                {'account_id': self.output_account.id, 'debit': 0, 'credit': 0, 'reconciled': False},
+            ],
+        })
+
+        # Set a serial number for the product
+        lot = self.env['stock.production.lot'].create({
+            'name': 'lot1',
+            'product_id': self.product1.id,
+            'company_id': self.env.company.id,
+        })
+
+        self.pos_session.picking_ids[0].move_line_ids[0].write({
+            'lot_id': lot.id
+        })
```

Comment:
if there is only one picking and one move_line, the `[0]` aren't needed.
The `[0]` makes me think there are more and there are, then we should explain why we only take the first one.

Reply:
If there are multiple for whatever reason I don't want it to write to multiple. Without being explicit about the `[0]` it's saying to write to the entire recordset, which aren't the semantics that I want to communicate. In `test_selling_products_with_a_partially_invalid_picking` there are multiple `move_line_ids`. What I could do is add some extra asserts on the lengths of the lists so it's explicit which lengths I'm expecting. I'd still keep the `[0]` to make clear I'm writing to a single record (or the first in the sorted list). What do you think about that?

Reply:
True, I forgot you could write on all of them at the same time. :man_facepalming: 

Fair enough, add a check on the length so it's explicit.
BTW, you can directly write `self.pos_session.picking_ids[0].move_line_ids[0].lot_id = lot`

Reply:
Ok, will add the length check. Regarding the write: I've been traumatized by server actions not allowing attribute assignments :-P. A bit more seriously: at a later point in the same code I write more values than just the `lot_id` (`qty_done`), where a `write` is cleaner, so I just stuck to the same convention for all of them.

Reply:
I'm adding a length check for `move_line_ids`, I just remembered why I didn't have one for `picking_ids`: because I set `'point_of_sale_update_stock_quantities': 'closing'` in the test setup, it's only going to create a single picking after closing the session, so verifying the length seemed redundant.